### DISCUSSION TITLE
ECR Repository and Proxy without Authentication Fails

### DISCUSSION THREAD

**Discussion by chezsmithy** [renovatebot/renovate#21710] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Tested on latest release and latest code from GitHub.

### Please select which platform you are using if self-hosting.

gitlab.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

15.1.2-ee

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When an ECR repository is configured with a friendly proxy, and that proxy doesn't require authentication lookups fail for tags as the 405 error is returned on the initial authentication call. Further, when the request for tags occurs a 405 is returned but without an HttpError so the code to switch to using 1000 vs 10000 tags is not triggered.

In the case of the 405 on the first call to getAuthHeaders, the debug logs indicate "Failed to get authHeaders for getTags lookup" due to the null response in the case of the 405.

Debug logs for the second 405 without error provide no real value in this situation as the code in index.ts getDockerApiTags isn't instrumented in this case so it fails silently. Additionally its been difficult to create a Jest test case for this scenario as Jest Mocks appear to always result in a 405 returning HttpError. However, in my live environment I am observing a 405 without an HttpError.

### Relevant debug logs

<details><summary>Logs</summary>

```
Debug log content noted in the description...
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment]
i don't see a bug here, as 405 is clearly an error on registry side

**Comment by chezsmithy** [comment]
I have opened a PR to demonstrate a fix that works: https://github.com/renovatebot/renovate/pull/16822

**Comment by chezsmithy** [comment]
@viceice for an ECR registry with tags set to 10000 a 405 is an indicator that the number of tags being requested is too large and we need to back off the requested number of tags to 1000.

**Comment by chezsmithy** [comment]
Here is the debug output from the relevant calls:

Entering into getAuthHeaders from getDockerApiTags from get Tags.

apiCheckUrl = 'https://docker.artifactory.cambiahealth.com/v2/chp-build-python-jobrunner/tags/list?n=10000'

Note .endswith('/v2') is false, so we make the await http.getJson(apiCheckUrl, options) call.

apiCheckReponse

```
{statusCode: 405, body: {…}, headers: {…}, authorization: false}
authorization:
false
body:
{errors: Array(1)}
errors:
(1) [{…}]
0:
{code: 'UNSUPPORTED', message: 'Invalid parameter at 'maxResults' failed to … must have value less than or equal to 1000''}
code:
'UNSUPPORTED'
message:
'Invalid parameter at 'maxResults' failed to satisfy constraint: 'Member must have value less than or equal to 1000''
__proto__:
Object
__proto__:
Array(0)
__proto__:
Object
length:
1
__proto__:
Object
headers:
{date: 'Thu, 28 Jul 2022 03:12:15 GMT', content-type: 'application/json; charset=utf-8', content-length: '164', connection: 'close', docker-distribution-api-version: 'registry/2.0', …}
statusCode: 405
```

This is expected as the proxy in-front of ECR does not require authentication so we don't hit the condition on line 50...

```
if (apiCheckResponse.statusCode !== 401 ||
            !is_1.default.nonEmptyString(apiCheckResponse.headers['www-authenticate'])) {
            logger_1.logger.warn({ apiCheckUrl, res: apiCheckResponse }, 'Invalid registry response');
            return null;
        }

```
Note that the 405 response was not an HttpError. I believe this is due to the fact that the request was made without headers. With headers it would fail as an HttpError. In our case we have configured a hostRule for this repo as:

```
 {
      hostType: 'docker',
      matchHost: 'docker.artifactory.cambiahealth.com',
      username: '',
      password: '',
      authType: 'Basic'
    },
```

This forces it to not attempt an ECR based IAM login.

When the code returns to getDockerApiTags headers are set.

It then proceeds to obtain tags with maxResults of 10000. This fails again, without Auth Headers but also without an HttpError thus bypassing the block of code which was previously added.

Thus we hit the added if statement, detect the 405 via the isECRMaxResultsResponse function and reset the maxResults to 1000. We continue, and get tags.

**Comment by chezsmithy** [comment]
@viceice I added additional  context above related to the program flow I am observing with out proxy.

**Reply by hhomar** [reply]
@chezsmithy I'm running into the same problem. I can see that the PR was closed, are you intending to do anymore work on the issue or have you found a workaround?

**Comment by firke** [comment]
Hitting this issue now as well, we have proxy fronting ECR, and it handles authentication itself.
This is the relevant log:
```
 WARN: Invalid registry response (repository=example)
       "apiCheckUrl": "https://example.com/v2/prod/docker.io/library/docker/tags/list?n=10000",
       "res": {
         "statusCode": 405,
         "body": {
           "errors": [
             {
               "code": "UNSUPPORTED",
               "message": "Invalid parameter at 'maxResults' failed to satisfy constraint: 'Member must have value less than or equal to 1000'"
             }
           ]
         },
         "headers": {
           "date": "Fri, 05 Jan 2024 09:50:08 GMT",
           "content-type": "application/json; charset=utf-8",
           "content-length": "164",
           "connection": "close",
           "docker-distribution-api-version": "registry/2.0",
           "sizes": ""
         },
         "authorization": false
       }
DEBUG: Failed to get authHeaders for getTags lookup (repository=example-repo)
```

So essentially as far as I understand, a request is made to getAuthHeaders, response is then 405 which is expected from an ECR, since it's max is 1000, but then it's returns null here:
```
    if (
      apiCheckResponse.statusCode !== 401 ||
      !is.nonEmptyString(apiCheckResponse.headers['www-authenticate'])
    ) {
      logger.warn(
        { apiCheckUrl, res: apiCheckResponse },
        'Invalid registry response',
      );
      return null;
    }
```

Which then fails here:
```
    if (!headers) {
      logger.debug('Failed to get authHeaders for getTags lookup');
      return null;
    }
```

imo there are two options, not sure which is better:
1. make count of tags configurable by env var, so we can specify a 1000 in the beginning
2. make a back off, e.g. if list tags is called and error is 405 + specific text, it will then backoff and try 1000 instead, this is already largely implemented, but could be moved up a bit?

Thank you!

**Reply by rarkins** [reply]
Option 2 is best. PRs welcome

**Reply by firke** [reply]
Hi @rarkins, I've raised a PR for this issue - https://github.com/renovatebot/renovate/pull/26618
Let me know if anything can be improved on it, or if there is a better patterns/ways to implement the same. Mostly however issue stems from the fact that we don't want to return null when there are no auth headers, as well as making sure they are not {}, since in this case error 405 will not be caught in 'catch'. Thanks!
